src/main/java/com/wanderersoftherift/wotr/block/QuestHubBlock.java

```diff
@@ -0,0 +1,105 @@
+
+    public QuestHubBlock(BlockBehaviour.Properties properties) {
+        super(properties);
+    }
+
+    protected @NotNull InteractionResult useWithoutItem(
+            @NotNull BlockState state,
+            Level level,
+            @NotNull BlockPos pos,
+            @NotNull Player player,
+            @NotNull BlockHitResult hitResult) {
+        if (level.isClientSide || !(player instanceof ServerPlayer serverPlayer)) {
+            return InteractionResult.SUCCESS;
+        }
+
+        ActiveQuests activeQuests = player.getData(WotrAttachments.ACTIVE_QUESTS);
+        if (activeQuests.isEmpty()) {
+            List<QuestState> availableQuests = getAvailableQuests(level, serverPlayer);
+
+            player.openMenu(
+                    new SimpleMenuProvider(
+                            (containerId, playerInventory, p) -> new QuestGiverMenu(containerId, playerInventory,
+                                    ContainerLevelAccess.create(level, pos), new ArrayList<>(availableQuests)),
+                            CONTAINER_TITLE)
+            );
+            PacketDistributor.sendToPlayer(serverPlayer, new AvailableQuestsPayload(availableQuests));
+        } else {
+            player.openMenu(
+                    new SimpleMenuProvider(
+                            (containerId, playerInventory, p) -> new QuestCompletionMenu(containerId, playerInventory,
+                                    ContainerLevelAccess.create(level, pos), activeQuests,
+                                    activeQuests.getQuestList().getFirst().getId()),
+                            Component.empty())
+            );
+        }
+        return InteractionResult.CONSUME;
+    }
+
+    private static @NotNull List<QuestState> getAvailableQuests(Level level, ServerPlayer serverPlayer) {
+        List<QuestState> availableQuests = serverPlayer.getData(WotrAttachments.AVAILABLE_QUESTS);
+        if (serverPlayer.getData(WotrAttachments.AVAILABLE_QUESTS).isEmpty()) {
+            availableQuests = generateNewQuestList(level, serverPlayer);
+            serverPlayer.setData(WotrAttachments.AVAILABLE_QUESTS, availableQuests);
+        }
+        return availableQuests;
+    }
+
+    private static @NotNull List<QuestState> generateNewQuestList(Level level, ServerPlayer serverPlayer) {
+        LootParams params = new LootParams.Builder(serverPlayer.serverLevel()).create(LootContextParamSets.EMPTY);
+        Registry<Quest> questRegistry = level.registryAccess().lookupOrThrow(WotrRegistries.Keys.QUESTS);
+        List<Quest> quests = questRegistry.stream().collect(Collectors.toList());
+        List<QuestState> generatedQuests = new ArrayList<>();
+        for (int i = 0; i < QUEST_SELECTION_SIZE && !quests.isEmpty(); i++) {
+            int index = level.random.nextInt(quests.size());
+            Quest quest = quests.removeLast();
+            if (index < quests.size() - 1) {
+                quest = quests.set(index, quests.removeLast());
+            }
```

Comment:
This is not equivalent to what it was before nor it matches my suggestion. 

I my original suggestion was this:
```suggestion
            Quest quest = quests.removeLast();
            if (index < quests.size()) {
                quest = quests.set(index, quest);
            }
```